core/include/kernel/dt.h

```diff
@@ -52,8 +52,15 @@ struct dt_device_match {
 	const char *compatible;
 };
 
+enum dt_driver_type {
+	DT_DRIVER_NOTYPE,
+	DT_DRIVER_UART,
+	DT_DRIVER_CLK,
+};
+
 struct dt_driver {
 	const char *name;
+	enum dt_driver_type type;
```

Comment:
It is not clear to me why this new field is needed. I can see how it can optimize the traversal of the driver list, but aren't the `compatible` strings enough to uniquely match a driver?

Reply:
The problem is that driver are found using the following:

```
	dt_drv = dt_find_compatible_driver(fdt, offs);
	if (!dt_drv || dt_drv->type != DT_DRIVER_UART)
		goto out;

	sdrv = (const struct serial_driver *)dt_drv->driver;
	if (!sdrv)
		goto out;
```
As you may see, the `driver` field is cast into a "serial/clock/whatever" specific structure. So, if the node compatible for a clock point to a serial driver for instance, we'll cast it into the wrong type and probably cause havoc.

Reply:
Ah, got it, sorry I read the old code a bit too quickly. Indeed this could be a problem in case the code is fed with a bad DT.
Could you please clarify in the commit description of "drivers: add type field to dt_driver struct"? My proposal: "...which will allow differentiating drivers when iterating on entries **and thus avoid casting struct dt_driver::driver to the wrong type in case a bad DT is provided**".

Reply:
Yep exactly, wrong DT could cause cast error (I agree that this is very unlikely but better be safe than sorry). Ok for the commit description.